contracts/compound-rate-keeper/AbstractCompoundRateKeeper.sol

```diff
@@ -0,0 +1,170 @@
+        try this.getCompoundRate() returns (uint256 rate_) {
+            if (rate_ == _getMaxRate()) {
+                isMaxRateReached = true;
+            }
+        } catch {
+            isMaxRateReached = true;
+        }
+    }
+
+    /**
+     *  @notice The function to get current compound rate
+     *  @return current compound rate
+     */
+    function getCompoundRate() public view override returns (uint256) {
+        return getFutureCompoundRate(uint64(block.timestamp));
+    }
+
+    /**
+     *  @notice The function to get future compound rate (the timestamp_ may be equal to the lastUpdate)
+     *  @param timestamp_ the timestamp to calculate the rate for
+     *  @return the compound rate for the provided timestamp
+     */
+    function getFutureCompoundRate(uint64 timestamp_) public view override returns (uint256) {
+        if (isMaxRateReached) {
+            return _getMaxRate();
+        }
+
+        uint64 lastUpdate_ = lastUpdate;
+
+        if (lastUpdate_ >= timestamp_) {
+            return _currentRate;
+        }
+
+        uint64 secondsPassed_ = timestamp_ - lastUpdate_;
+
+        uint64 capitalizationPeriod_ = capitalizationPeriod;
+        uint64 capitalizationPeriodsNum_ = secondsPassed_ / capitalizationPeriod_;
+        uint64 secondsLeft_ = secondsPassed_ % capitalizationPeriod_;
+
+        uint256 capitalizationRate_ = capitalizationRate;
+        uint256 rate_ = _currentRate;
+
+        if (capitalizationPeriodsNum_ != 0) {
+            uint256 capitalizationPeriodRate_ = capitalizationRate_.rpow(
+                capitalizationPeriodsNum_,
+                PRECISION
+            );
+            rate_ = (rate_ * capitalizationPeriodRate_) / PRECISION;
+        }
+
+        if (secondsLeft_ > 0) {
+            uint256 rateLeft_ = PRECISION +
+                ((capitalizationRate_ - PRECISION) * secondsLeft_) /
+                capitalizationPeriod_;
+            rate_ = (rate_ * rateLeft_) / PRECISION;
+        }
+
+        return rate_.min(_getMaxRate());
+    }
+
+    /**
+     *  @notice The internal function to set the capitalization rate
+     *  @param capitalizationRate_ new capitalization rate
+     */
+    function _setCapitalizationRate(uint256 capitalizationRate_) internal {
+        require(!isMaxRateReached, "CRK: max rate is reached");
+
+        _currentRate = getCompoundRate();
+        lastUpdate = uint64(block.timestamp);
+
+        _changeCapitalizationRate(capitalizationRate_);
+    }
+
+    /**
+     *  @notice The internal function to set the capitalization period
+     *  @param capitalizationPeriod_ new capitalization period
+     */
+    function _setCapitalizationPeriod(uint64 capitalizationPeriod_) internal {
+        require(!isMaxRateReached, "CRK: max rate is reached");
+
+        _currentRate = getCompoundRate();
+        lastUpdate = uint64(block.timestamp);
+
+        _changeCapitalizationPeriod(capitalizationPeriod_);
+    }
```

Comment:
We can create an internal function to avoid code duplication. This function can take a pointer to a function to achieve the necessary logic